frontend/javascripts/oxalis/model/sagas/min_cut_saga.js

```diff
@@ -0,0 +1,694 @@
+    return;
+  }
+
+  const segmentId = inputData[ll(seedA)];
+
+  // We calculate the latest time at which the min cut may still
+  // be executed. The min-cut algorithm will check from time to
+  // time whether the threshold is violated.
+  // This approach was chosen over a timeout mechanism via
+  // race/delay saga due to performance reasons. The latter approach
+  // requires that even tightly-called functions are generators
+  // which yield to the redux-saga middleware. Rough benchmarks
+  // showed an overall slow-down of factor 6.
+  const timeoutThreshold = performance.now() + MIN_CUT_TIMEOUT;
+  console.log("Starting min-cut...");
+  console.time("Total min-cut");
+
+  console.time("Build Graph");
+
+  const edgeBuffer = buildGraph(
+    inputData,
+    segmentId,
+    size,
+    boundingBoxTarget.getVolume(),
+    l,
+    ll,
+    timeoutThreshold,
+  );
+
+  // Copy original edge buffer, as it's mutated later when removing edges.
+  const originalEdgeBuffer = new Uint16Array(edgeBuffer);
+  console.timeEnd("Build Graph");
+
+  console.time("Find & delete paths");
+
+  let exitLoop = false;
+  while (!exitLoop) {
+    const { foundTarget, distanceField, directionField } = populateDistanceField(
+      edgeBuffer,
+      boundingBoxTarget,
+      seedA,
+      seedB,
+      ll,
+      timeoutThreshold,
+    );
+    if (foundTarget) {
+      const { removedEdgeCount } = removeShortestPath(
+        distanceField,
+        directionField,
+        seedA,
+        seedB,
+        ll,
+        size,
+        edgeBuffer,
+        minDistToSeed,
+      );
+      if (removedEdgeCount === 0) {
+        throw PartitionFailedError;
+      }
+    } else {
+      console.log("Segmentation is partitioned");
+      exitLoop = true;
+    }
+  }
+
+  console.timeEnd("Find & delete paths");
+
+  console.time("traverseResidualsField");
+  const { visitedField } = traverseResidualsField(boundingBoxTarget, seedA, ll, edgeBuffer);
+  console.timeEnd("traverseResidualsField");
+
+  console.time("labelDeletedEdges");
+  labelDeletedEdges(visitedField, boundingBoxTarget, size, originalEdgeBuffer, targetMag, l, ll);
+  console.timeEnd("labelDeletedEdges");
+
+  console.timeEnd("Total min-cut");
+
+  console.log({ seedA, seedB, boundingBoxMag1, inputData, edgeBuffer });
+}
+
+function isPositionOutside(position, size) {
+  return (
+    position[0] < 0 ||
+    position[1] < 0 ||
+    position[2] < 0 ||
+    position[0] >= size[0] ||
+    position[1] >= size[1] ||
+    position[2] >= size[2]
+  );
+}
+
+function buildGraph(inputData, segmentId, size, length, l, ll, timeoutThreshold) {
+  const edgeBuffer = new Uint16Array(length);
+  for (let x = 0; x < size[0]; x++) {
+    if (x % Math.floor(size[0] / 10) === 0 && performance.now() > timeoutThreshold) {
+      // After each 10% chunk, we yield to redux-saga to allow
+      // cancellation.
+      throw TimeoutError;
+    }
+    for (let y = 0; y < size[1]; y++) {
+      for (let z = 0; z < size[2]; z++) {
```

Comment:
should I reorder the x-y-z loops to z-y-x for better cache locality? probably no harm in changing it.